### PR TITLE
[REVIEW] Moving cuml sparse prims to raft

### PR DESCRIPTION
**Pull request by cjnolet**
Moving cuml sparse prims and gtests to raft. The namespaces have already been adjusted in cuml so this is a simple move for the most part. There are a few places where includes needed to be updated and I Needed to remove cuml debug logs.

### CODE REVIEW

**Comment by cjnolet** [comment]
Also tagging @lowener because recent changes to brute_force_knn (both dense and sparse) will also need to be moved into here.

**Comment by cjnolet** [comment]
rerun tests

File: cpp/include/raft/sparse/convert/dense.cuh [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+#include <iostream>
+
+#include <raft/sparse/utils.h>
+
+namespace raft {
+namespace sparse {
+namespace convert {
+
+template <typename value_t>
+__global__ void csr_to_dense_warp_per_row_kernel(int n_cols,
```
Comment (divyegala) [inline cpp/include/raft/sparse/convert/dense.cuh:40]: Will there be a future need to template these ints?

File: cpp/include/raft/sparse/convert/csr.cuh [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+#include <raft/sparse/coo.cuh>
+#include <raft/sparse/linalg/degree.cuh>
+#include <raft/sparse/op/row_op.cuh>
+
+namespace raft {
+namespace sparse {
+namespace convert {
+
+template <typename value_t>
+void coo_to_csr(const raft::handle_t &handle, const int *srcRows,
```
Comment (divyegala) [inline cpp/include/raft/sparse/convert/csr.cuh:47]: template on ints?

File: cpp/include/raft/sparse/convert/csr.cuh [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+ *
+ * @param rows: COO rows array
+ * @param nnz: size of COO rows array
+ * @param row_ind: output row indices array
+ * @param m: number of rows in dense matrix
+ * @param d_alloc device allocator for temporary buffers
+ * @param stream: cuda stream to use
+ */
+template <typename T>
+void sorted_coo_to_csr(const T *rows, int nnz, T *row_ind, int m,
```
Comment (divyegala) [inline cpp/include/raft/sparse/convert/csr.cuh:154]: Template on ints

File: cpp/include/raft/sparse/convert/csr.cuh [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+/**
+ * @brief Generate the row indices array for a sorted COO matrix
+ *
+ * @param coo: Input COO matrix
+ * @param row_ind: output row indices array
+ * @param d_alloc device allocator for temporary buffers
+ * @param stream: cuda stream to use
+ */
+template <typename T>
+void sorted_coo_to_csr(COO<T> *coo, int *row_ind,
```
Comment (divyegala) [inline cpp/include/raft/sparse/convert/csr.cuh:180]: Template on ints

File: cpp/include/raft/sparse/coo.cuh [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+    return out;
+  }
+
+  /**
+    * @brief Set the number of rows and cols
+    * @param n_rows: number of rows in the dense matrix
+    * @param n_cols: number of columns in the dense matrix
+    */
+  void setSize(int n_rows, int n_cols) {
```
Comment (divyegala) [inline cpp/include/raft/sparse/coo.cuh:201]: This should pick up `Index_Type`

File: cpp/include/raft/sparse/coo.cuh [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    this->n_rows = n_rows;
+    this->n_cols = n_cols;
+  }
+
+  /**
+    * @brief Set the number of rows and cols for a square dense matrix
+    * @param n: number of rows and cols
+    */
+  void setSize(int n) {
```
Comment (divyegala) [inline cpp/include/raft/sparse/coo.cuh:210]: Same as above

File: cpp/include/raft/sparse/coo.cuh [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    this->n_cols = n;
+  }
+
+  /**
+    * @brief Allocate the underlying arrays
+    * @param nnz: size of underlying row/col/val arrays
+    * @param init: should values be initialized to 0?
+    * @param stream: CUDA stream to use
+    */
+  void allocate(int nnz, bool init, cudaStream_t stream) {
```
Comment (divyegala) [inline cpp/include/raft/sparse/coo.cuh:221]: Same as above

File: cpp/include/raft/sparse/coo.cuh [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+  }
+
+  /**
+    * @brief Allocate the underlying arrays
+    * @param nnz: size of the underlying row/col/val arrays
+    * @param size: the number of rows/cols in a square dense matrix
+    * @param init: should values be initialized to 0?
+    * @param stream: CUDA stream to use
+    */
+  void allocate(int nnz, int size, bool init, cudaStream_t stream) {
```
Comment (divyegala) [inline cpp/include/raft/sparse/coo.cuh:232]: Same as above

File: cpp/include/raft/sparse/coo.cuh [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+  /**
+    * @brief Allocate the underlying arrays
+    * @param nnz: size of the underlying row/col/val arrays
+    * @param n_rows: number of rows in the dense matrix
+    * @param n_cols: number of columns in the dense matrix
+    * @param init: should values be initialized to 0?
+    * @param stream: stream to use for init
+    */
+  void allocate(int nnz, int n_rows, int n_cols, bool init,
```
Comment (divyegala) [inline cpp/include/raft/sparse/coo.cuh:244]: Same as above

File: cpp/include/raft/sparse/linalg/norm.cuh [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+namespace raft {
+namespace sparse {
+namespace linalg {
+
+template <int TPB_X = 64, typename T>
+__global__ void csr_row_normalize_l1_kernel(
+  // @TODO: This can be done much more parallel by
+  // having threads in a warp compute the sum in parallel
+  // over each row and then divide the values in parallel.
+  const int *ia,           // csr row ex_scan (sorted by row)
```
Comment (divyegala) [inline cpp/include/raft/sparse/linalg/norm.cuh:44]: Template on int

File: cpp/include/raft/sparse/linalg/norm.cuh [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+ *
+ * @param ia: row_ind array
+ * @param vals: data array
+ * @param nnz: size of data array
+ * @param m: size of row_ind array
+ * @param result: l1 normalized data array
+ * @param stream: cuda stream to use
+ */
+template <int TPB_X = 64, typename T>
+void csr_row_normalize_l1(const int *ia,  // csr row ex_scan (sorted by row)
```
Comment (divyegala) [inline cpp/include/raft/sparse/linalg/norm.cuh:88]: Template on int

File: cpp/include/raft/sparse/linalg/norm.cuh [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    <<<grid, blk, 0, stream>>>(ia, vals, nnz, m, result);
+  CUDA_CHECK(cudaGetLastError());
+}
+
+template <int TPB_X = 64, typename T>
+__global__ void csr_row_normalize_max_kernel(
+  // @TODO: This can be done much more parallel by
+  // having threads in a warp compute the sum in parallel
+  // over each row and then divide the values in parallel.
+  const int *ia,           // csr row ind array (sorted by row)
```
Comment (divyegala) [inline cpp/include/raft/sparse/linalg/norm.cuh:108]: Template on int

File: cpp/include/raft/sparse/linalg/norm.cuh [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+ * @param ia: row_ind array
+ * @param vals: data array
+ * @param nnz: size of data array
+ * @param m: size of row_ind array
+ * @param result: l1 normalized data array
+ * @param stream: cuda stream to use
+ */
+
+template <int TPB_X = 64, typename T>
+void csr_row_normalize_max(const int *ia,  // csr row ind array (sorted by row)
```
Comment (divyegala) [inline cpp/include/raft/sparse/linalg/norm.cuh:154]: Template on int

File: cpp/include/raft/sparse/op/filter.cuh [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+#include <raft/sparse/utils.h>
+#include <raft/sparse/coo.cuh>
+#include <raft/sparse/linalg/degree.cuh>
+
+namespace raft {
+namespace sparse {
+namespace op {
+
+template <int TPB_X, typename T>
+__global__ void coo_remove_zeros_kernel(const int *rows, const int *cols,
```
Comment (divyegala) [inline cpp/include/raft/sparse/op/filter.cuh:45]: Template on int

File: cpp/include/raft/sparse/op/filter.cuh [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+        ccols[cur_out_idx] = cols[idx];
+        cvals[cur_out_idx] = vals[idx];
+        ++cur_out_idx;
+      }
+    }
+  }
+}
+
+template <int TPB_X, typename T>
+__global__ void coo_remove_scalar_kernel(const int *rows, const int *cols,
```
Comment (divyegala) [inline cpp/include/raft/sparse/op/filter.cuh:68]: Template on int

File: cpp/include/raft/sparse/op/filter.cuh [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+ * @param cvals: compressed array of vals
+ * @param cnnz: array of non-zero counts per row
+ * @param cur_cnnz array of counts per row
+ * @param scalar: scalar to remove from arrays
+ * @param n: number of rows in dense matrix
+ * @param d_alloc device allocator for temporary buffers
+ * @param stream: cuda stream to use
+ */
+template <int TPB_X, typename T>
+void coo_remove_scalar(const int *rows, const int *cols, const T *vals, int nnz,
```
Comment (divyegala) [inline cpp/include/raft/sparse/op/filter.cuh:108]: Template on int

File: cpp/include/raft/sparse/op/sort.h [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+ * @param n number of cols in coo matrix
+ * @param nnz number of non-zeros
+ * @param rows rows array from coo matrix
+ * @param cols cols array from coo matrix
+ * @param vals vals array from coo matrix
+ * @param d_alloc device allocator for temporary buffers
+ * @param stream: cuda stream to use
+ */
+template <typename T>
+void coo_sort(int m, int n, int nnz, int *rows, int *cols, T *vals,
```
Comment (divyegala) [inline cpp/include/raft/sparse/op/sort.h:57]: Template on int

**Review by divyegala — APPROVED** [review]
Changes LGTM to go! Just minor comments about templating that would be nice to track in an issue

**Comment by cjnolet** [comment]
Thanks @divyegala! Created https://github.com/rapidsai/raft/issues/147 to track the template changes. Will merge this as-is for now just to remove the deltas.

**Comment by cjnolet** [comment]
@gpucibot merge

**Comment by cjnolet** [comment]
@gpucibot merge

**Review by dantegd — APPROVED** [review]

**Pull request merged**